src/mbgl/gl/attributes.hpp

```diff
@@ -0,0 +1,154 @@
+template <> struct AttributeType<int16_t>  : std::integral_constant<GLenum, GL_SHORT> {};
+template <> struct AttributeType<uint16_t> : std::integral_constant<GLenum, GL_UNSIGNED_SHORT> {};
+template <> struct AttributeType<int32_t>  : std::integral_constant<GLenum, GL_INT> {};
+template <> struct AttributeType<uint32_t> : std::integral_constant<GLenum, GL_UNSIGNED_INT> {};
+template <> struct AttributeType<float>    : std::integral_constant<GLenum, GL_FLOAT> {};
+template <> struct AttributeType<double>   : std::integral_constant<GLenum, GL_DOUBLE> {};
+
+template <GLint N, class T>
+class VertexAttribute {
+public:
+    GLuint index;
+    ByteOffset byteOffset;
+    ByteLength byteStride;
+};
+
+class VertexBuffer {
+public:
+    using Attribute = variant<
+        VertexAttribute<1, int8_t>,
+        VertexAttribute<2, int8_t>,
+        VertexAttribute<3, int8_t>,
+        VertexAttribute<4, int8_t>,
+        VertexAttribute<1, uint8_t>,
+        VertexAttribute<2, uint8_t>,
+        VertexAttribute<3, uint8_t>,
+        VertexAttribute<4, uint8_t>,
+        VertexAttribute<1, int16_t>,
+        VertexAttribute<2, int16_t>,
+        VertexAttribute<3, int16_t>,
+        VertexAttribute<4, int16_t>,
+        VertexAttribute<1, uint16_t>,
+        VertexAttribute<2, uint16_t>,
+        VertexAttribute<3, uint16_t>,
+        VertexAttribute<4, uint16_t>,
+        VertexAttribute<1, int32_t>,
+        VertexAttribute<2, int32_t>,
+        VertexAttribute<3, int32_t>,
+        VertexAttribute<4, int32_t>,
+        VertexAttribute<1, uint32_t>,
+        VertexAttribute<2, uint32_t>,
+        VertexAttribute<3, uint32_t>,
+        VertexAttribute<4, uint32_t>,
+        VertexAttribute<1, float>,
+        VertexAttribute<2, float>,
+        VertexAttribute<3, float>,
+        VertexAttribute<4, float>,
+        VertexAttribute<1, double>,
+        VertexAttribute<2, double>,
+        VertexAttribute<3, double>,
+        VertexAttribute<4, double>>;
+
+    UniqueBuffer buffer;
+    std::vector<Attribute> attributes;
+};
+
+class VertexBufferSlice {
+public:
+    const VertexBuffer& buffer;
+    ByteOffset byteOffset;
+    ByteLength byteLength;
+};
+
+template <class T>
+class ConstantScalar {
+public:
+    GLuint index;
+    T value;
+};
+
+template <GLint N, class T>
+class ConstantVector {
+public:
+    GLuint index;
+    std::array<T, N> value;
+};
+
+using AttributeSource = variant<
+    VertexBufferSlice,
+    ConstantScalar<int16_t>,
+    ConstantScalar<float>,
+    ConstantScalar<double>
+    /*
+    ConstantVector<2, int16_t>,
+    ConstantVector<3, int16_t>,
+    ConstantVector<4, int16_t>,
+    ConstantVector<2, float>,
+    ConstantVector<3, float>,
+    ConstantVector<4, float>,
+    ConstantVector<2, double>,
+    ConstantVector<3, double>,
+    ConstantVector<4, double>
+    */
+    >;
+
+using AttributeSources = std::vector<AttributeSource>;
+
+namespace detail {
```

Comment:
Or did you just mean move them to an `_impl.hpp` file? I assumed you meant move them to a `.cpp` file.